ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/BlockManager.java

```diff
@@ -124,87 +133,119 @@ private void importBlockIgnoringResult(final SignedBeaconBlock block) {
   }
 
   private SafeFuture<BlockImportResult> doImportBlock(final SignedBeaconBlock block) {
-    notifyReceivedBlockSubscribers(block);
-    if (!shouldImportBlock(block)) {
-      return SafeFuture.completedFuture(BlockImportResult.knownBlock(block));
-    }
+    final Optional<BlockImportResult> blockImportResult =
+        handleInvalidBlock(block)
+            .or(
+                () -> {
+                  notifyReceivedBlockSubscribers(block);
+                  return handleChildOfFailedExecutionPayloadBlock(block);
+                })
+            .or(() -> handleKnownBlock(block));
+    return blockImportResult
+        .map(SafeFuture::completedFuture)
+        .orElseGet(
```

Comment:
I think I can improve here by returning `Optional<SafeFuture>` in `handle*` methods.